substrate/frame/transaction-payment/asset-conversion-tx-payment/src/payment.rs

```diff
@@ -49,150 +53,219 @@ pub trait OnChargeAssetTransaction<T: Config> {
+			}
+		} else {
+			// Check if the refund amount can be swapped back into the asset used by `who` for fee
+			// payment.
+			let refund_asset_amount = S::quote_price_exact_tokens_for_tokens(
+				A::get(),
+				asset_id.clone(),
+				refund_amount,
+				true,
 			)
-			.ok()
-			{
-				Some(acquired) => {
-					asset_refund = acquired
-						.try_into()
-						.map_err(|_| TransactionValidityError::from(InvalidTransaction::Payment))?;
-				},
-				None => {
-					Pallet::<T>::deposit_event(Event::<T>::AssetRefundFailed {
-						native_amount_kept: swap_back,
-					});
-				},
+			// No refund given if it cannot be swapped back.
+			.unwrap_or(Zero::zero());
+
+			// Deposit the refund before the swap to ensure it can be processed.
+			let debt = match F::deposit(
+				asset_id.clone(),
+				&who,
+				refund_asset_amount,
+				Precision::BestEffort,
+			) {
+				Ok(debt) => debt,
+				// No refund given since it cannot be deposited.
+				Err(_) => fungibles::Debt::<T::AccountId, F>::zero(asset_id.clone()),
+			};
+
+			if debt.peek().is_zero() {
+				// No refund given.
+				(initial_asset_consumed, fee_paid)
+			} else {
+				let (refund, fee_paid) = fee_paid.split(refund_amount);
+				match S::swap_exact_tokens_for_tokens(
+					vec![A::get(), asset_id],
+					refund,
+					Some(refund_asset_amount),
+				) {
+					Ok(refund_asset) => {
+						match refund_asset.offset(debt) {
+							Ok(SameOrOther::None) => {},
+							// This arm should never be reached, as the  amount of `debt` is
+							// expected to be exactly equal to the amount of `refund_asset` credit.
+							_ => {
+								defensive!("Debt should be equal to the refund credit");
+								return Err(InvalidTransaction::Payment.into())
+							},
+						};
+						(
+							initial_asset_consumed.saturating_sub(refund_asset_amount.into()),
+							fee_paid,
+						)
+					},
+					// The error should not occur since swap was quoted before.
+					Err((refund, _)) => {
+						defensive!("Refund swap should pass for the quoted amount");
+						match F::settle(who, debt, Preservation::Expendable) {
+							Ok(dust) => ensure!(dust.peek().is_zero(), InvalidTransaction::Payment),
+							// The error should not occur as the `debt` was just withdrawn above.
+							Err(_) => {
+								defensive!("Should settle the debt");
+								return Err(InvalidTransaction::Payment.into())
+							},
+						};
+						let fee_paid = fee_paid.merge(refund).map_err(|_| {
+							// The error should never occur since `fee_paid` and `refund` are
+							// credits of the same asset.
+							InvalidTransaction::Payment
+						})?;
+						(initial_asset_consumed, fee_paid)
+					},
+				}
 			}
-		}
+		};
 
-		let actual_paid = initial_asset_consumed.saturating_sub(asset_refund);
-		Ok(actual_paid)
+		// Handle the imbalance (fee and tip separately).
+		let (tip, fee) = adjusted_paid.split(tip);
+		OUF::on_unbalanced(fee);
+		OUT::on_unbalanced(tip);
```

Comment:
we can make this as in transaction payment pallet (`OU::on_unbalanceds(vec![fee, tip])`), if https://github.com/paritytech/polkadot-sdk/pull/4564 pr get merged